File: src/ui/workspace.rs
use crate::app::state::AppState;
use crate::ui::converter_selector::show_converter_selector;
use crate::ui::param_panel::show_param_panel;
use crate::ui::plot_panel::show_plot_panel;
use crate::ui::result_panel::show_result_panel;
use crate::ui::schematic_view::show_schematic;

/// Render the main workspace layout with all panels.
pub fn show_workspace(ui: &mut egui::Ui, state: &mut AppState) {
    // Left panel: parameters
    egui::Panel::left("param_panel")
        .resizable(true)
        .default_size(250.0)
        .min_size(200.0)
        .show_inside(ui, |ui| {
            egui::ScrollArea::vertical().show(ui, |ui| {
                ui.heading("⚙ Parameters");
                ui.separator();
                show_converter_selector(ui, state);
                ui.add_space(8.0);
                show_param_panel(ui, state);
            });
        });

    // Right panel: results
    egui::Panel::right("result_panel")
        .resizable(true)
        .default_size(220.0)
        .min_size(180.0)
        .show_inside(ui, |ui| {
            egui::ScrollArea::vertical().show(ui, |ui| {
                show_result_panel(ui, state);
            });
        });

    // Center panel: schematic + plots
    egui::CentralPanel::default().show_inside(ui, |ui| {
        // Vertical split: schematic on top, plots on bottom
        egui::Panel::top("schematic_area")
            .resizable(true)
            .default_size(180.0)
            .min_size(120.0)
            .show_inside(ui, |ui| {
                ui.heading("🔌 Schematic");
                ui.separator();
                show_schematic(ui, state);
            });

        // Bottom: plots
        egui::CentralPanel::default().show_inside(ui, |ui| {
            ui.heading("📈 Waveforms");
            ui.separator();
            show_plot_panel(ui, state);
        });
    });
}

File: tupan/ui/workspace.py
"""Main workspace layout with 3-panel QSplitter.

Left: Design params + computed components
Center: Schematic + Waveform plots
Right: Results (operating point + performance)
"""

from PySide6.QtCore import Qt
from PySide6.QtWidgets import (
    QWidget, QVBoxLayout, QHBoxLayout, QSplitter, QScrollArea,
    QGroupBox,
)

from tupan.app.state import AppState
from tupan.ui.converter_selector import ConverterSelector
from tupan.ui.param_panel import ParamPanel
from tupan.ui.result_panel import ResultPanel
from tupan.ui.schematic_view import SchematicView
from tupan.ui.plot_panel import PlotPanel


class WorkspaceWidget(QWidget):

    def __init__(self, state, parent=None):
        super().__init__(parent)
        self.state = state
        self._setup_ui()

    def _setup_ui(self):
        layout = QHBoxLayout(self)
        layout.setContentsMargins(0, 0, 0, 0)

        splitter = QSplitter(Qt.Orientation.Horizontal)

        # ── Left panel: Parameters ──
        left = QWidget()
        left_layout = QVBoxLayout(left)
        left_layout.setContentsMargins(8, 8, 8, 8)

        self.converter_selector = ConverterSelector(self.state)
        self.converter_selector.converter_changed.connect(
            self._on_converter_changed
        )
        left_layout.addWidget(self.converter_selector)

        left_layout.addSpacing(8)

        self.param_panel = ParamPanel(self.state)
        self.param_panel.params_changed.connect(self._on_params_changed)
        left_layout.addWidget(self.param_panel)

        left_layout.addStretch()
        splitter.addWidget(left)

        # ── Center panel: Schematic + Plots ──
        center = QWidget()
        center_layout = QVBoxLayout(center)
        center_layout.setContentsMargins(8, 8, 8, 8)

        sg = QGroupBox("Schematic")
        sgl = QVBoxLayout(sg)
        self.schematic_view = SchematicView(self.state)
        sgl.addWidget(self.schematic_view)
        center_layout.addWidget(sg)

        pg = QGroupBox("Waveforms")
        pgl = QVBoxLayout(pg)
        self.plot_panel = PlotPanel(self.state)
        pgl.addWidget(self.plot_panel)
        center_layout.addWidget(pg)

        splitter.addWidget(center)

        # ── Right panel: Results ──
        right_scroll = QScrollArea()
        right_scroll.setWidgetResizable(True)
        right_scroll.setHorizontalScrollBarPolicy(
            Qt.ScrollBarPolicy.ScrollBarAlwaysOff
        )
        self.result_panel = ResultPanel(self.state)
        right_scroll.setWidget(self.result_panel)
        splitter.addWidget(right_scroll)

        splitter.setSizes([320, 480, 220])
        splitter.setStretchFactor(0, 0)
        splitter.setStretchFactor(1, 1)
        splitter.setStretchFactor(2, 0)

        layout.addWidget(splitter)

    def _on_converter_changed(self, ct):
        self.param_panel.update_state(self.state)
        self.result_panel.update_state(self.state)
        self.schematic_view.update_state(self.state)
        self.plot_panel.update_state(self.state)

    def _on_params_changed(self):
        """Design parameter changed → recalculate already happened in param_panel."""
        self.result_panel.update_state(self.state)
        self.schematic_view.update_state(self.state)
        self.plot_panel.update_state(self.state)

    def refresh_plot_theme(self):
        self.plot_panel.refresh_theme()

    def update_state(self, state):
        self.state = state
        self.converter_selector.update_state(state)
        self.param_panel.update_state(state)
        self.result_panel.update_state(state)
        self.schematic_view.update_state(state)
        self.plot_panel.update_state(state)
